File: frontend/src/api/planner.ts
/**
 * Meal Planner API client
 * Type-safe functions for all planner-related API calls
 */

import type { PlanRequest, PlanResponse } from '../types/planner'

const API_BASE = '/api/v1/plan'

class ApiError extends Error {
  constructor(
    public status: number,
    message: string
  ) {
    super(message)
    this.name = 'ApiError'
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}))
    throw new ApiError(
      response.status,
      errorData.detail || `HTTP ${response.status}`
    )
  }
  
  return response.json()
}

export const plannerApi = {
  /**
   * Generate a 7-day meal plan based on preferences
   */
  async generatePlan(request: PlanRequest = {}): Promise<PlanResponse> {
    const response = await fetch(`${API_BASE}/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    })
    return handleResponse<PlanResponse>(response)
  },

  /**
   * Get all available cuisine tags for filtering
   */
  async getCuisines(): Promise<string[]> {
    const response = await fetch(`${API_BASE}/cuisines`)
    return handleResponse<string[]>(response)
  },

  /**
   * Get all available diet tags for filtering
   */
  async getDiets(): Promise<string[]> {
    const response = await fetch(`${API_BASE}/diets`)
    return handleResponse<string[]>(response)
  },
}

export { ApiError }

File: backend/app/api/routes/planner.py
"""
Meal Planner API routes.

Endpoints:
- POST /plan/generate  - Generate a 7-day meal plan
- GET  /plan/cuisines  - Get available cuisine tags
- GET  /plan/diets     - Get available diet tags
"""

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.crud.recipe import get_recipes, get_all_cuisines, get_all_diets
from app.services.planner import generate_meal_plan
from app.schemas.planner import PlanRequest, PlanResponse


router = APIRouter(
    prefix="/plan",
    tags=["Meal Planner"],
)


# =============================================================================
# POST /plan/generate - Generate meal plan
# =============================================================================

@router.post(
    "/generate",
    response_model=PlanResponse,
    status_code=status.HTTP_200_OK,
    summary="Generate meal plan",
    description="Generate a personalized 7-day meal plan based on preferences and constraints.",
)
async def generate_plan(
    request: PlanRequest,
    db: AsyncSession = Depends(get_db),
) -> PlanResponse:
    """
    Generate a 7-day meal plan.
    
    **Request Parameters:**
    - **budget**: Max weekly budget in $ (optional)
    - **max_time**: Max cooking time per meal in minutes (default: 60)
    - **protein_goal**: Target protein per meal in grams (optional)
    - **diet_tags**: Required dietary tags like "vegetarian" (optional)
    - **cuisine_preferences**: Preferred cuisines like "italian" (optional)
    - **servings_per_day**: Meals per day, 1-3 (default: 1)
    
    **Response:**
    - 7-day meal plan with recipes
    - Aggregated shopping list
    - Summary statistics (cost, protein, calories)
    
    **Example Request:**
    ```json
    {
        "budget": 75,
        "max_time": 45,
        "protein_goal": 30,
        "diet_tags": ["gluten-free"],
        "cuisine_preferences": ["italian", "mexican"]
    }
    ```
    """
    # Fetch all recipes from database
    recipes, total = await get_recipes(db, limit=1000)
    
    if total == 0:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="No recipes available. Please try again later.",
        )
    
    # Generate the meal plan
    plan = generate_meal_plan(recipes, request)
    
    return plan


# =============================================================================
# GET /plan/cuisines - Get available cuisines
# =============================================================================

@router.get(
    "/cuisines",
    response_model=list[str],
    summary="Get available cuisines",
    description="Get all unique cuisine tags from the recipe database.",
)
async def list_cuisines(
    db: AsyncSession = Depends(get_db),
) -> list[str]:
    """
    Get available cuisine tags for filtering.
    
    Returns a sorted list of unique cuisines like:
    `["american", "asian", "italian", "mexican", ...]`
    
    Use these values in the `cuisine_preferences` field when generating a plan.
    """
    cuisines = await get_all_cuisines(db)
    return cuisines


# =============================================================================
# GET /plan/diets - Get available diet tags
# =============================================================================

@router.get(
    "/diets",
    response_model=list[str],
    summary="Get available diet tags",
    description="Get all unique diet tags from the recipe database.",
)
async def list_diets(
    db: AsyncSession = Depends(get_db),
) -> list[str]:
    """
    Get available diet tags for filtering.
    
    Returns a sorted list of unique diet tags like:
    `["gluten-free", "high-protein", "vegetarian", "vegan", ...]`
    
    Use these values in the `diet_tags` field when generating a plan.
    """
    diets = await get_all_diets(db)
    return diets
